adding service integrations to service overview row

philbrookes opened the pull request:
It is possible for various mobile services to be bound to each other, these changes to the UI will make it easier for a mobile developer using OpenShift to host the services for their mobile client to connect these services to each other.

Creating an integration between services will create a binding between the two services and also create a pod preset for the binding values to be connected to the correct pods. The deployment of the consuming service will then be updated to trigger a fresh deployment with the new config values present.

There are various states possible:
1. A service is running which can integrate with a service that is not running:
![integrations-provision](https://user-images.githubusercontent.com/1033452/38039083-6399a588-32a4-11e8-96c8-1577a3c85f39.png)

1.1 A service is being provisioned, which can be integrated with:
![integrations-provisioning](https://user-images.githubusercontent.com/1033452/38039106-7183fc20-32a4-11e8-8e19-44189dfff658.png)

2. A service is running which can integrate with a service that is also running:
![integrations-no-binding](https://user-images.githubusercontent.com/1033452/38039119-78d71200-32a4-11e8-8744-8ba02286529f.png)

3. A service is in the process of being integrated or de-integrated with another service:
![integrations-pending-binding](https://user-images.githubusercontent.com/1033452/38039128-7e22ff3a-32a4-11e8-826a-1c6305d87f34.png)

4. A service is currently integrated with another service:
![integrations-binding-active](https://user-images.githubusercontent.com/1033452/38039149-84d2178a-32a4-11e8-97a0-a37aecd065cd.png)

4. A service which could be integrated with, is being deprovisioned:
![integrations-service-deprovisioning](https://user-images.githubusercontent.com/1033452/38039175-93b2cb78-32a4-11e8-8db6-93908b110440.png)

spadgett commented:
/ok-to-test
cc @cshinn

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+        ctrl.binding = _.filter(data, function(binding){
+          var bindingProviderName = _.get(binding, "metadata.annotations.provider", "provider_not_found");
+          var bindingConsumerName = _.get(binding, "metadata.annotations.consumer", "consumer_not_found");
+          var consumerName = _.get(ctrl, "consumerService.metadata.labels.serviceName", "consumer_name_not_found");
```
I'd rather just check for null / undefined instead of using something like "provider_not_found".

spadgett replied:
We typically namespace annotations. Something like `servicecatalog.openshift.io/consumer`.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+      //setup watch on servincebindings, watch for bindings consumed by this service
+      DataService.watch(bindingPreferredVersion, context, function(bindingData) {
+        var data = bindingData.by("metadata.name");
+        ctrl.binding = _.filter(data, function(binding){
```
Use `_.find` instead of `_.filter` / `pop`

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+      //setup watch on service instances, looking for an instance which provides this integration
+      DataService.watch(instancePreferredVersion, context, function(serviceInstancesData) {
+        var data = serviceInstancesData.by('metadata.name');
+        ctrl.serviceInstance = _.filter(data, function(serviceInstance){
```
_.find

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+    ctrl.$onInit = function() {
+      var context = {namespace: ctrl.consumerService.metadata.namespace};
+      // watch for the pod preset for this integration
+      DataService.watch(podPresetPreferredVersion, context, function(podPresets){
```
You're leaking websockets since you're not calling `DataService.unwatch` on these watches.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+      var consumerSvcName = _.get(consumerService, 'metadata.name');
+      var providerSvcName = _.get(providerService, 'metadata.name');
+      var podPreset = {
+        "apiVersion": "settings.k8s.io/v1alpha1",
```
nit: Since the keys here and below don't need to be quoted

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+    var deploymentPreferredVersion = APIService.getPreferredVersion('deployments');
+    var bindingPreferredVersion = APIService.getPreferredVersion('servicebindings');
+    var instancePreferredVersion = APIService.getPreferredVersion('serviceinstances');
+    var podPresetPreferredVersion = {group: "settings.k8s.io", version: "v1alpha1", resource: "podpresets"};
```
We've also made a statement not to use alpha APIs in the console. We need to be forward-compatible with an n+1 API server for upgrades when masters can be at different versions.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+          DataService.unwatch(bindingReadyWatch);
+          NotificationsService.addNotification({
+            type: "success",
+            message: "Binding has been set in " + _.get(ctrl, 'consumerService.metadata.labels.serviceName') + " Redeployed " + _.get(ctrl, 'consumerService.metadata.labels.serviceName')
```
We try to use complete sentences with punctuation in these messages.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+            NotificationsService.addNotification({
+              type: "error",
+              message: "Failed to integrate service binding",
+              details: err.data.message
```
getErrorDetails filter

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+          .catch(function(err) {
+            NotificationsService.addNotification({
+              type: "error",
+              message: "Failed to integrate service binding",
```
Add punctuation.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+        );
+      })
+      .then(function(deployment) {
+        delete deployment.spec.template.metadata.labels[integrationName];
```
Use `angular.copy` before changing the deployment. DataService does some caching and you don't want to modify the cached data.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+      .catch(function(error){
+        NotificationsService.addNotification({
+          type: "error",
+          message: "error removing integration",
```
Should be a complete sentence.

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+        NotificationsService.addNotification({
+          type: "error",
+          message: "error removing integration",
+          details: error.data.message
```
getErrorDetails filter

spadgett commented on app/scripts/filters/mobile.js:
```diff
@@ -0,0 +1,9 @@
+angular.module('openshiftConsole')
+  .filter('isMobileService', function() {
+    return function(serviceInstance) {
+      return _.get(serviceInstance, 'metadata.labels', {}).mobile === 'enabled';
```
```js
return _.get(serviceInstance, 'metadata.labels.mobile') === 'enabled';
```

spadgett commented on app/scripts/directives/serviceIntegration.js:
```diff
@@ -0,0 +1,242 @@
+            type: "success",
+            message: "Binding has been set in " + _.get(ctrl, 'consumerService.metadata.labels.serviceName') + " Redeployed " + _.get(ctrl, 'consumerService.metadata.labels.serviceName')
+          });
+          _.set(watchBinding, "metadata.annotations.consumer", ctrl.consumerService.metadata.labels.serviceName);
```
Use `angular.copy` before modifying the binding.

spadgett commented on app/scripts/directives/overview/serviceInstanceRow.js:
```diff
@@ -75,6 +79,12 @@
       if (changes.bindings) {
         row.deleteableBindings = _.reject(row.bindings, 'metadata.deletionTimestamp');
       }
+      if(row.isMobileService(row.apiObject) && $rootScope.AEROGEAR_MOBILE_ENABLED){
```
nit: spaces before `(` and `{`, here and below

spadgett commented on app/scripts/directives/overview/serviceInstanceRow.js:
```diff
@@ -75,6 +79,12 @@
         row.deleteableBindings = _.reject(row.bindings, 'metadata.deletionTimestamp');
       }
+      if(row.isMobileService(row.apiObject) && $rootScope.AEROGEAR_MOBILE_ENABLED){
+        var integrations = _.get(getServiceClass(), "spec.externalMetadata.integrations", false);
```
You don't need to give a default here since undefined is falsey. If you do, I'd rather it be `''` since we're expecting a string, but better to leave it off.

```js
var serviceClass = getServiceClass();
var integrations = _.get(serviceClass, "spec.externalMetadata.integrations");
```

spadgett commented:
/assign

cshinn commented:
@philbrookes I'm having Visuals take a look at it, and I'll get back to you ASAP

cshinn commented:
@philbrookes Here are a few small suggestions.
![group 3](https://user-images.githubusercontent.com/3300756/37660958-04757bc0-2c2a-11e8-8a2b-37ed7b7b84b1.png)

philbrookes commented:
@spadgett @cshinn I have implemented the changes as you have suggested - as much as I could make them work.

Some thoughts:
- The podpreset needs to stay as a watch, rather than a watchObject as it can be deleted and recreated and the watchobject does not appear to handle that correctly.
- The pull-left does seem to be required, if I remove that the UI looks broken.
- The podpreset version is pulled from the APIService now, rather than hard-coding the API version.
- I've added a new state which reflects that the integratable service is being provisioned.

Let me know if you have any thoughts or questions.

philbrookes commented:
@spadgett  @cshinn I have rebased this now, and retested against a 3.9.0 cluster, can I get a fresh review please?

spadgett commented:
/retest

spadgett commented:
> I have rebased this now, and retested against a 3.9.0 cluster, can I get a fresh review please?

Sorry following up on this:

* We're past feature freeze for 3.10 (for a while now), so we'll need to wait until master opens for 3.11
* We'll need to look at the state of pod presets in 3.11 before we can add that alpha API. Alpha APIs cause a lot of rework and can break things during upgrades when masters are at different versions since alpha APIs are not stable
